searx/engines/brave.py

```diff
@@ -0,0 +1,250 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
+"""
+ Brave (General, news, videos, images)
+"""
+# pylint: disable=fixme
+
+from urllib.parse import (
+    urlencode,
+    urlparse,
+    parse_qs,
+)
+
+import chompjs
+from lxml import html
+
+from searx.utils import (
+    extract_text,
+    eval_xpath_list,
+    eval_xpath_getindex,
+)
+
+about = {
+    "website": 'https://search.brave.com/',
+    "wikidata_id": 'Q22906900',
+    "official_api_documentation": None,
+    "use_official_api": False,
+    "require_api_key": False,
+    "results": 'HTML',
+}
+
+base_url = "https://search.brave.com/"
+categories = []
+brave_category = 'search'  # search, images, videos, news
+brave_spellcheck = False
+
+send_accept_language_header = True
+paging = False
+"""By Brave, paging is only supported in :py:obj:`brave_category`
+``search`` (UI category All)."""
+
+safesearch = True
+safesearch_map = {2: 'strict', 1: 'moderate', 0: 'off'}  # cookie: safesearch=off
+
+time_range_support = False
+"""By Brave, time range search is only supported in :py:obj:`brave_category`
+``search`` (UI category All)."""
+
+time_range_map = {
+    'day': 'pd',
+    'week': 'pw',
+    'month': 'pm',
+    'year': 'py',
+}
+
+
+def request(query, params):
+
+    # Don't accept br encoding / see https://github.com/searxng/searxng/pull/1787
+    params['headers']['Accept-Encoding'] = 'gzip, deflate'
+
+    args = {
+        'q': query,
+    }
+    if brave_spellcheck:
+        args['spellcheck'] = '1'
+
+    if brave_category == 'search':
+        if params.get('pageno', 1) - 1:
+            args['offset'] = params.get('pageno', 1) - 1
+        if time_range_map.get(params['time_range']):
+            args['tf'] = time_range_map.get(params['time_range'])
+
+    params["url"] = f"{base_url}{brave_category}?{urlencode(args)}"
+
+    # set preferences in cookie
+    params['cookies']['safesearch'] = safesearch_map.get(params['safesearch'], 'off')
+
+    # ToDo: we need a fetch_traits(..) implementation / the ui_lang of Brave are
+    #       limited and the country handling has it quirks
+
+    eng_locale = params.get('searxng_locale')
+    params['cookies']['useLocation'] = '0'  # the useLocation is IP based, we use 'country'
+
+    if not eng_locale or eng_locale == 'all':
+        params['cookies']['country'] = 'all'  # country=all
+    else:
+        params['cookies']['country'] = eng_locale.split('-')[-1].lower()
+        params['cookies']['ui_lang'] = eng_locale.split('-')[0].lower()
+
+    # logger.debug("cookies %s", params['cookies'])
+
+
+def response(resp):
+
+    if brave_category == 'search':
+        return _parse_search(resp)
+
+    datastr = ""
+    for line in resp.text.split("\n"):
+        if "const data = " in line:
```

Comment:
maybe we should save "const data = " to a variable called `data_identifier` or something similar since it's used twice

Reply:
I don't think this is need, in this line its used to find a position in a string while the user usage replaces this string with an empty string ... Since this implementation depends on line breaks and strings being found in certain places, it cannot be ruled out that we will have to do some rework here soon (e.g. if the line breaks are set differently one day than they are today).